adaptor.go

```diff
@@ -0,0 +1,138 @@
+package jwtverifier
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/lestrrat-go/jwx/jws"
+	"github.com/patrickmn/go-cache"
+)
+
+type Adaptor interface {
+	Decode(jwt string) (map[string]interface{}, error)
+}
+
+type JWXAdaptor struct {
+	jwksUri     string
+	jwkKeyCache *cache.Cache
+}
+
+func NewJWXAdaptor(metaDataUrl string) (Adaptor, error) {
+	jwksUri, err := getJWKSURI(metaDataUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed go get jwks uri: %w", err)
+	}
+
+	return &JWXAdaptor{
+		jwksUri:     jwksUri,
+		jwkKeyCache: cache.New(24*time.Hour, time.Hour),
+	}, nil
+}
+
+func (a *JWXAdaptor) Decode(jwt string) (map[string]interface{}, error) {
+	kid, err := getKid(jwt)
+	if err != nil {
+		return nil, err
+	}
+
+	jwkKey, err := a.getJwkKey(kid)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := jws.VerifyWithJWK([]byte(jwt), jwkKey)
+	if err != nil {
+		return nil, err
+	}
+
+	var claims map[string]interface{}
+	if err = json.Unmarshal(token, &claims); err != nil {
+		return claims, err
+	}
+
+	return claims, nil
+
+}
+
+func (a *JWXAdaptor) getJwkKey(kid string) (jwk.Key, error) {
+	if x, found := a.jwkKeyCache.Get(kid); found {
+		switch v := x.(type) {
+		case error:
+			return nil, v
+		case jwk.Key:
+			return v, nil
+		default:
+			return nil, errors.New("unknown value found in jwkKeyCache")
+		}
+	}
+
+	jwkSet, err := jwk.FetchHTTP(a.jwksUri)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: This probably needs to be just the key...
+	keys := jwkSet.LookupKeyID(kid)
+	if len(keys) < 1 {
+		a.jwkKeyCache.SetDefault(kid, ErrJWKNotFound)
+		return nil, ErrJWKNotFound
+	}
+
+	key := keys[len(keys)-1]
+	a.jwkKeyCache.SetDefault(kid, key)
+
+	return key, nil
+}
+
+func getJWKSURI(metaDataUrl string) (string, error) {
```

Comment:
I'm curious why it has been done like this in the first place. The `/keys` endpoint is well documented https://developer.okta.com/docs/reference/api/oidc/#keys, the rules for constructing `${baseUrl}` are known ahead of time https://developer.okta.com/docs/reference/api/oidc/#composing-your-base-url. So one could just construct the `jwks_uri` without making any external calls, I think.

Anyway it's better than before because we make only one request during initialization and we don't need  https://github.com/okta/okta-jwt-verifier-golang/blob/bf6f0d73000e3873d519714a3619b3c5d62ba615/jwtverifier.go#L36 anymore

Reply:
This is how it is normally implemented in an openid client out of convenience. As long as you know the metadata url, then you can configure the whole client with that. Probably not necessary here if we know that that url is never going to change.